handlers/mtn/mtn.go

```diff
@@ -0,0 +1,245 @@
+	payload := &moPayload{}
+
+	err := handlers.DecodeAndValidateJSON(payload, r)
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	date := time.Unix(payload.Created/1000, payload.Created%1000*1000000).UTC()
+	urn, err := handlers.StrictTelForCountry(payload.From, channel.Country())
+	if err != nil {
+		return nil, handlers.WriteAndLogRequestError(ctx, h, channel, w, r, err)
+	}
+
+	// create our message
+	msg := h.Backend().NewIncomingMsg(channel, urn, payload.Message, clog).WithReceivedOn(date)
+	// and finally write our message
+	return handlers.WriteMsgsAndResponse(ctx, h, []courier.Msg{msg}, w, r, clog)
+}
+
+type mtPayload struct {
+	From    string   `json:"senderAddress"`
+	To      []string `json:"receiverAddress"`
+	Message string   `json:"message"`
+}
+
+// Send implements courier.ChannelHandler
+func (h *handler) Send(ctx context.Context, msg courier.Msg, clog *courier.ChannelLog) (courier.MsgStatus, error) {
+	accessToken, err := h.getAccessToken(ctx, msg.Channel(), clog)
+	if err != nil {
+		return nil, err
+	}
+
+	partSendURL, _ := url.Parse(sendURL)
+
+	status := h.Backend().NewMsgStatusForID(msg.Channel(), msg.ID(), courier.MsgErrored, clog)
+	parts := handlers.SplitMsgByChannel(msg.Channel(), handlers.GetTextAndAttachments(msg), maxMsgLength)
+	for i, part := range parts {
+		mtMsg := &mtPayload{}
+		mtMsg.From = strings.TrimPrefix(msg.Channel().Address(), "+")
+		mtMsg.To = []string{strings.TrimPrefix(msg.URN().Path(), "+")}
+		mtMsg.Message = part
+
+		requestBody := &bytes.Buffer{}
+		json.NewEncoder(requestBody).Encode(mtMsg)
+
+		// build our request
+		req, err := http.NewRequest(http.MethodPost, partSendURL.String(), requestBody)
+		if err != nil {
+			return nil, err
+		}
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("Accept", "application/json")
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", accessToken))
+
+		resp, respBody, err := handlers.RequestHTTP(req, clog)
+		if err != nil || resp.StatusCode/100 != 2 {
+			return status, nil
+		}
+
+		externalID, err := jsonparser.GetString(respBody, "transactionId")
+		if err != nil {
+			clog.Error(courier.ErrorResponseValueMissing("transactionId"))
+			return status, nil
+		}
+
+		// if this is our first message, record the external id
+		if i == 0 {
+			status.SetExternalID(externalID)
+			status.SetStatus(courier.MsgWired)
+		}
+
+	}
+
+	return status, nil
+}
+
+func (h *handler) RedactValues(ch courier.Channel) []string {
+	return []string{
+		ch.StringConfigForKey(courier.ConfigAPIKey, ""),
+		ch.StringConfigForKey(courier.ConfigAuthToken, ""),
+	}
+}
+
+func (h *handler) getAccessToken(ctx context.Context, channel courier.Channel, clog *courier.ChannelLog) (string, error) {
+	rc := h.Backend().RedisPool().Get()
+	defer rc.Close()
+
+	tokenKey := fmt.Sprintf("channel-token:%s", channel.UUID())
+
+	h.fetchTokenMutex.Lock()
+	defer h.fetchTokenMutex.Unlock()
```

Comment:
I know this is how it's implemented for the other channels like this.. but I remember thinking this is potentially a race condition across different boxes (depending on how MTN API works). 2 different courier boxes could get in here at the same time and fetch new access tokens at the same time.. 

If requesting the second one invalidates the first one.. then there's a problem. If both tokens are valid then there's no problem. Not a huge problem either way because even if the first token is now invalid, we fail to send that mesage, but we'll get the second token when we try again.

All that to say we can merge it like this, but we should look into this.